modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C

```diff
@@ -89,10 +90,15 @@ CohesiveZoneActionBase::validParams()
 
 CohesiveZoneActionBase::CohesiveZoneActionBase(const InputParameters & params) : Action(params)
 {
+  // FIXME: suggest to use action of action to add this to avoid changing the input parameters in
+  // the warehouse.
+  const auto & parameters = _app.getInputParameterWarehouse().getInputParameters();
+  InputParameters & pars(*(parameters.find(uniqueActionName())->second.get()));
```

Comment:
Why can't this be in the base class constructor? And why do the parameters in the warehouse have to be modifiable for anyone this to work?

Reply:
and what is "action of action"?

Reply:
The base class can construct a non-const reference. Not get the second question.

"action of action" means a meta action that can call `ActionFactory::create` to add a new action. In there, the action can populate necessary parameters of the action to be created and then create the action. The action to be created will not register any input syntax. Creating the action can happen at `meta_action` task which is at the beginning of execution of all actions.

Reply:
@dschwen are you ok with this or you prefer me to add the non-const function to get the action parameters in the base class?

Reply:
I don't want to see
```c++
InputParameters & parameters();
```
added back to `Action`. I would rather you just `const_cast<InputParameters &>(_pars)` here because a `const_cast` is an indication that a developer is doing something that we'd rather not have them do. That you can do the above two lines and get a non-const reference to the parameters is unfortunate IMO and I don't want it advertised.

Reply:
ok.

Reply:
@dschwen can we resolve this?